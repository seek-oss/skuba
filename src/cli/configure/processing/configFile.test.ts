import {
  generateIgnoreFileSimpleVariants,
  generateNpmrcSimpleVariants,
  mergeWithConfigFile,
} from './configFile';

describe('generateIgnoreFileSimpleVariants', () => {
  it.each([
    ['variant path', ['/lib*/'], ['/lib*/', '/lib/', '/lib', 'lib/', 'lib']],
    ['non-variant path', ['lib'], ['/lib', '/lib/', 'lib', 'lib/']],
    ['duplicate patterns', ['lib', 'lib'], ['/lib', '/lib/', 'lib', 'lib/']],
    [
      'file extension',
      ['*.tgz'],
      ['*.tgz', '.tgz', '.tgz/', '/.tgz', '/.tgz/'],
    ],
    ['potential empty string', ['/'], ['/']],
    ['empty string', [''], []],
  ])('handles %s', (_, pattern, expected) =>
    expect(generateIgnoreFileSimpleVariants(pattern)).toEqual(
      new Set(expected),
    ),
  );
});

describe('generateNpmrcSimpleVariants', () => {
  it.each([
    [
      'quotes added',
      [`public-hoist-pattern[]="tsconfig-seek"`],
      [
        `public-hoist-pattern[]="tsconfig-seek"`,
        `public-hoist-pattern[]=tsconfig-seek`,
      ],
    ],
    [
      'quotes removed',
      [`public-hoist-pattern[]=tsconfig-seek`],
      [
        `public-hoist-pattern[]="tsconfig-seek"`,
        `public-hoist-pattern[]=tsconfig-seek`,
      ],
    ],
    ['weird line is ignored', ['a = b = c'], ['a = b = c']],
    ['empty string', [''], []],
  ])('handles %s', (_, pattern, expected) =>
    expect(generateNpmrcSimpleVariants(pattern)).toEqual(new Set(expected)),
  );
});

describe('mergeWithConfigFile for ignore files', () => {
  const baseTemplate =
    '# managed by skuba\nnode_modules*/\n# end managed by skuba\n';
  const updatedBaseTemplate =
    '# managed by skuba\nnode_modules*/\n.DS_Store\n# end managed by skuba\n';

  const cases = [
    ['empty provided', baseTemplate, ''],

    ['provided with no managed section', baseTemplate, '.DS_Store\n'],
    [
      'provided with managed section and partially superseded config',
      updatedBaseTemplate,
      'system32\n\n.DS_Store\nnode_modules/\n\n*.zip\n',
    ],
    [
      'provided with managed section and fully superseded config',
      updatedBaseTemplate,
      '\r\n\nnode_modules\r\nnode_modules_bak\n/node_modules',
    ],
    [
      'provided with outdated managed section',
      updatedBaseTemplate,
      baseTemplate,
    ],
    ['identical', baseTemplate, baseTemplate],
    [
      'provided with managed section and additional lines',
      baseTemplate,
      `.idea\n.vscode\n\n${baseTemplate}\n.DS_Store\nnode_modules\n`,
    ],
    [
      'provided with outdated managed section and additional lines',
      updatedBaseTemplate,
      `.idea\n.vscode\n\n${baseTemplate}\n.DS_Store\nnode_modules\n`,
    ],
  ] as const;

  test.each(cases)('%s', (_, base, provided) =>
    expect(mergeWithConfigFile(base)(provided)).toMatchSnapshot(),
  );

  test.each(cases)('%s', (_, base, provided) =>
    expect(mergeWithConfigFile(base, 'npmrc')(provided)).toMatchSnapshot(),
  );

  it('produces stable output over multiple runs', () => {
    const merge = mergeWithConfigFile(baseTemplate);

    let input = baseTemplate;

    for (let i = 0; i < 5; i++) {
      input = merge(input);

      expect(input).toBe(baseTemplate);
    }
  });
});

describe('mergeWithConfigFile for npmrc files', () => {
  const baseTemplate =
    '# managed by skuba\npublic-hoist-pattern[]="tsconfig-seek"\n# end managed by skuba\n';
  const updatedBaseTemplate =
    '# managed by skuba\npublic-hoist-pattern[]="tsconfig-seek"\npublic-hoist-pattern[]="jest"\n# end managed by skuba\n';

  const cases = [
    ['empty provided', baseTemplate, ''],

    [
      'provided with no managed section',
      baseTemplate,
      'public-hoist-pattern[]="mystuff"',
    ],
    [
      'provided with managed section and partially superseded config',
      updatedBaseTemplate,
      'public-hoist-pattern[]=a\npublic-hoist-pattern[]=jest',
    ],
    [
      'provided with managed section and fully superseded config',
      updatedBaseTemplate,
      'public-hoist-pattern[]="jest"',
    ],
    [
      'provided with outdated managed section',
      updatedBaseTemplate,
      baseTemplate,
    ],
    ['identical', baseTemplate, baseTemplate],
    [
      'provided with managed section and additional lines',
      baseTemplate,
      `public-hoist-pattern[]=a\n\n${baseTemplate}\npublic-hoist-pattern[]=b`,
    ],
    [
      'provided with outdated managed section and additional lines',
      updatedBaseTemplate,
      `public-hoist-pattern[]=a\n\n${baseTemplate}\npublic-hoist-pattern[]="jest"=b\n`,
    ],
  ] as const;

  test.each(cases)('%s', (_, base, provided) =>
    expect(mergeWithConfigFile(base, 'npmrc')(provided)).toMatchSnapshot(),
  );

  it('produces stable output over multiple runs', () => {
    const merge = mergeWithConfigFile(baseTemplate, 'npmrc');

    let input = baseTemplate;

    for (let i = 0; i < 5; i++) {
      input = merge(input);

      expect(input).toBe(baseTemplate);
    }
  });
});
