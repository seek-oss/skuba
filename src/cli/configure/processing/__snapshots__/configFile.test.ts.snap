// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeWithConfigFile for ignore files empty provided 1`] = `
"# managed by skuba
node_modules*/
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files empty provided 2`] = `
"# managed by skuba
node_modules*/
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files identical 1`] = `
"# managed by skuba
node_modules*/
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files identical 2`] = `
"# managed by skuba
node_modules*/
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and additional lines 1`] = `
".idea
.vscode

# managed by skuba
node_modules*/
# end managed by skuba

.DS_Store
node_modules
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and additional lines 2`] = `
".idea
.vscode

# managed by skuba
node_modules*/
# end managed by skuba

.DS_Store
node_modules
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and fully superseded config 1`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and fully superseded config 2`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba

node_modules
node_modules_bak
/node_modules
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and partially superseded config 1`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba

system32

*.zip
"
`;

exports[`mergeWithConfigFile for ignore files provided with managed section and partially superseded config 2`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba

system32

node_modules/

*.zip
"
`;

exports[`mergeWithConfigFile for ignore files provided with no managed section 1`] = `
"# managed by skuba
node_modules*/
# end managed by skuba

.DS_Store
"
`;

exports[`mergeWithConfigFile for ignore files provided with no managed section 2`] = `
"# managed by skuba
node_modules*/
# end managed by skuba

.DS_Store
"
`;

exports[`mergeWithConfigFile for ignore files provided with outdated managed section 1`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files provided with outdated managed section 2`] = `
"# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for ignore files provided with outdated managed section and additional lines 1`] = `
".idea
.vscode

# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba

.DS_Store
node_modules
"
`;

exports[`mergeWithConfigFile for ignore files provided with outdated managed section and additional lines 2`] = `
".idea
.vscode

# managed by skuba
node_modules*/
.DS_Store
# end managed by skuba

.DS_Store
node_modules
"
`;

exports[`mergeWithConfigFile for npmrc files empty provided 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for npmrc files identical 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for npmrc files provided with managed section and additional lines 1`] = `
"public-hoist-pattern[]=a

# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
# end managed by skuba

public-hoist-pattern[]=b"
`;

exports[`mergeWithConfigFile for npmrc files provided with managed section and fully superseded config 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
public-hoist-pattern[]="jest"
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for npmrc files provided with managed section and partially superseded config 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
public-hoist-pattern[]="jest"
# end managed by skuba

public-hoist-pattern[]=a
"
`;

exports[`mergeWithConfigFile for npmrc files provided with no managed section 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
# end managed by skuba

public-hoist-pattern[]="mystuff"
"
`;

exports[`mergeWithConfigFile for npmrc files provided with outdated managed section 1`] = `
"# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
public-hoist-pattern[]="jest"
# end managed by skuba
"
`;

exports[`mergeWithConfigFile for npmrc files provided with outdated managed section and additional lines 1`] = `
"public-hoist-pattern[]=a

# managed by skuba
public-hoist-pattern[]="tsconfig-seek"
public-hoist-pattern[]="jest"
# end managed by skuba

public-hoist-pattern[]="jest"=b
"
`;
